### ISSUE TITLE
Negative Likelihood in OTF training

### ISSUE THREAD

**Issue by eric-yu-zhu** [mir-group/flare_pp#50]
Dear MIR team,

I got a large negative likelihood while running the otf training script PtH/pt_otf.py associated with your recent publication Nat Commun 13, 5183 (2022). The scripts were downloaded from https://archive.materialscloud.org/record/2022.92 

I tried flare (1.3.0) and flare_pp and both output a negative likelihood at the first step:

>> flare output
	GP hyperparameters: 
	Hyp0 : Hyp0                           = 2.2552
	Hyp1 : Hyp1                           = 0.0953
	Hyp2 : Hyp2                           = 0.3775
	Hyp3 : Hyp3                           = 0.0006
	Likelihood: -138.8188
	Likelihood gradient: [ 3.39102050e-01 -2.26372999e+00  4.38864081e+00 -2.38407634e+03]

>> flare_pp output
	GP hyperparameters: 
	Hyp0 : Hyp0                           = 4.0660
	Hyp1 : Hyp1                           = 0.4016
	Hyp2 : Hyp2                           = 0.4357
	Hyp3 : Hyp3                           = 0.0010
	Likelihood: -197.1289
	Likelihood gradient: [ 0.03468765  0.06725551 -0.04548742  0.49477924]

I had no problem running other otf training scripts including H2/h2_otf.py, Pt-bulk/pt_otf.py, Pt-surface/Pt_otf.py.

I also checked PtH/PtH.out and found that your likelihood values look very normal
    Likelihood: 89.3493 
    Likelihood: 388.1707 
    Likelihood: 660.0860 

What do you think the possible reasons for the discrepancy? How do I fix it?
 
May I know the flare_pp version (git commit) that produced PtH/PtH.out with PtH/pt_otf.py?

Best,
Eric

**Comment by YuuuXie** [comment]
Hi @eric-yu-zhu , it is normal that for some complicated system there could be negative log likelihood. In your case, when you observe different likelihood values from different trainings, possible reasons include different initializations, different hyperparameters etc. If you can run a few more steps, you can probably check if the likelihood goes up. 

If you want to check an old flare_pp version, I don't know the exact commit that was used for this training, but since the arXiv paper was submitted on 3rd Jun 2021, you can probably check the version right before that.

**Comment by eric-yu-zhu** [comment]
Thank you for the reply. Different random seeds and more md steps did not help unfortunately. The problem turns out to be caused by a subtle error on my side. 

What I learned from this and other previous otf training is that a negative likelihood can be a strong indicator that something is wrong.

Eric

**Comment by YuuuXie** [comment]
Hi @eric-yu-zhu , thank you for your clarification. May I ask what error was made to result in the negative likelihood? Just in case any other user could encounter the same issue in the future, it might be helpful to know

**Comment by eric-yu-zhu** [comment]
I guess the error is very specific to my project. But here are the details FYI.

The objective is to connect our dft engine to flare. So my first step is to write a vasp force calculator and reproduce the result generated by the VASP calculator imported from ase. My force calculator worked well for systems with single species such as bulk Pt and H2 molecules. For systems with multiple species, the problem is caused by the code writing POSCAR from atoms:

symbol_list = atoms.get_chemical_symbols()
element_list, count_list = np.unique(symbol_list, return_counts=True)

where np.unique reordered the elements according to the alphabet order, producing ['H', 'Pt']. On the other hand, the POTCAR was generated assuming the element order ['Pt', 'H']. 

Even worse, vasp detected the discrepancy but did not issue an error. Instead it changed ['H', 'Pt'] back to ['Pt, H'] but did not change the atom number accordingly. So I was running a very unphysical system, Pt(19)H(54), resulting in a large negative likelihood. With the bug fixed, the likelihood looks very normal now.

I still do not understand why the script did not work with flare_pp in which VASP calculator imported from ase was used. I'll try to do a debugging later this week and keep you posted if I find anything.

Best,
Eric

**Comment by YuuuXie** [comment]
Hi @eric-yu-zhu , thanks for the information. I'm not sure about how you set up the VASP calculator in ASE. But we also provide the yaml version for the on-the-fly training. So you could take a look. And the ASE VASP calculator is also applicable

**Comment by eric-yu-zhu** [comment]
An update on PtH2 otf training:

1. The inconsistency between two runs with flare and flare_pp was caused by a mistake in my script. After the correction, flare an flare_pp are consistent.

2. I got a vasp error while training PtH2: "Inconsistent Bravais lattice types found for crystalline and reciprocal lattice". I did experiments with ISYM, SYMPREC, and KPOINTS but could not resolve the issue. It turns out the problem is associated with VASP6. The problem is solved by downgrading to VASP5.
